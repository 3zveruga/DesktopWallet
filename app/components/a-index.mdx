---
name: Home
route: /
---

# UI Components

Use the sidebar menu to browse the styleguide of React components available for the Zepio application.
