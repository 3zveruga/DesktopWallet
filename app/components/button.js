// @flow

import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable max-len */
// $FlowFixMe
import { darken } from 'polished';

const DefaultButton = styled.button`
  padding: 10px 30px;
  font-family: ${// $FlowFixMe
  props => props.theme.fontFamily};
  font-weight: ${// $FlowFixMe
  props => props.theme.fontWeight.bold};
  font-size: ${// $FlowFixMe
  props => `${props.theme.fontSize.text}em`};
  cursor: pointer;
  outline: none;
  min-width: 100px;
  border-radius: 100px;
  transition: background-color 0.1s ease-in-out;
  width: 100%;
`;

const Primary = styled(DefaultButton)`
  background-color: ${props => props.theme.colors.primary};
  color: ${props => props.theme.colors.secondary};
  border: none;

  &:hover {
    background-color: ${props => darken(0.1, props.theme.colors.primary(props))};
  }

  &:disabled {
    background-color: #3e3c42;
    cursor: not-allowed;
    opacity: 0.8;
  }
`;

const Secondary = styled(DefaultButton)`
  background-color: transparent;
  color: ${props => props.theme.colors.secondary};
  border: 2px solid #3e3c42;

  &:hover {
    border-color: ${props => props.theme.colors.primary};
  }

  &:disabled {
    background-color: Transparent;
    cursor: not-allowed;
    color: #3e3c42;

    &:hover {
      border-color: #3e3c42;
    }
  }
`;

type Props = {
  label: string,
  onClick?: () => void,
  to?: string,
  variant?: 'primary' | 'secondary',
  disabled?: boolean,
  className?: string,
};

export const Button = ({
  onClick,
  label,
  to,
  variant,
  disabled,
  className,
}: Props) => {
  if (to && onClick) throw new Error('Should define either "to" or "onClick"');

  const component = variant === 'primary' ? (
    <Primary onClick={onClick} disabled={disabled} className={className}>
      {label}
    </Primary>
  ) : (
    <Secondary onClick={onClick} disabled={disabled} className={className}>
      {label}
    </Secondary>
  );

  return to ? <Link to={String(to)}>{component}</Link> : component;
};

Button.defaultProps = {
  to: null,
  variant: 'primary',
  onClick: null,
  disabled: false,
  className: '',
};
