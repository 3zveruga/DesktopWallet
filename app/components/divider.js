// @flow
import styled from 'styled-components';

export const Divider = styled.div`
  width: 100%;
  height: 1px;
  background-color: ${props => props.color || props.theme.colors.text};
  opacity: ${props => props.opacity || 1};
  margin-bottom: ${props => props.marginBottom || 0};
  margin-top: ${props => props.marginTop || 0};
`;
