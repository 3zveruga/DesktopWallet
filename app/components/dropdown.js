// @flow
import React, { type Node, Component } from 'react';
import styled from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
// $FlowFixMe
import { darken } from 'polished';
import Popover from 'react-popover';
import ClickOutside from 'react-click-outside';

import { TextComponent } from './text';

/* eslint-disable max-len  */
const MenuWrapper = styled.div`
  background-image: ${props => `linear-gradient(to right, ${darken(
    0.05,
    props.theme.colors.activeItem,
  )}, ${props.theme.colors.activeItem})`};
  padding: 10px 0;
  border-radius: 10px;
  margin-left: -10px;
`;

const MenuItem = styled.button`
  outline: none;
  background-color: transparent;
  border: none;
  border-bottom-style: solid;
  border-bottom-color: ${props => props.theme.colors.text};
  border-bottom-width: 1px;
  padding: 15px 10px;
  cursor: pointer;
  width: 100%;
  text-align: left;

  &:hover {
    opacity: 0.9;
  }

  &:disabled {
    cursor: default;

    &:hover {
      opacity: 1;
    }
  }
`;

const PopoverWithStyle = styled(Popover)`
  & > .Popover-tip {
    fill: ${props => props.theme.colors.activeItem};
  }
`;

type Props = {
  renderTrigger: (toggleVisibility: () => void, isOpen: boolean) => Node,
  options: Array<{ label: string, onClick: () => void }>,
  label?: string | null,
};

type State = {
  isOpen: boolean,
};

export class DropdownComponent extends Component<Props, State> {
  state = {
    isOpen: false,
  };

  static defaultProps = {
    label: null,
  };

  render() {
    const { isOpen } = this.state;
    const { label, options, renderTrigger } = this.props;

    return (
      <PopoverWithStyle
        isOpen={isOpen}
        preferPlace='below'
        enterExitTransitionDurationMs={0}
        body={[
          <ClickOutside
            onClickOutside={() => this.setState(() => ({ isOpen: false }))}
          >
            <MenuWrapper>
              {label && (
                <MenuItem disabled>
                  <TextComponent value={label} isBold />
                </MenuItem>
              )}
              {options.map(({ label: optionLabel, onClick }) => (
                <MenuItem onClick={onClick}>
                  <TextComponent value={optionLabel} />
                </MenuItem>
              ))}
            </MenuWrapper>
          </ClickOutside>,
        ]}
        tipSize={7}
      >
        {renderTrigger(
          () => this.setState(state => ({ isOpen: !state.isOpen })),
          isOpen,
        )}
      </PopoverWithStyle>
    );
  }
}
