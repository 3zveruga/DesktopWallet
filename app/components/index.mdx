---
name: Home
route: /
---

# UI Components

Use the sidebar menu to browse the styleguide
