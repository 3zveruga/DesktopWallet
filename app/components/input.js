// @flow
import React, { type Element } from 'react';
import styled from 'styled-components';

import theme from '../theme';

const getDefaultStyles = t => styled[t]`
  border-radius: ${props => props.theme.boxBorderRadius};
  border: none;
  background-color: ${props => props.bgColor || props.theme.colors.inputBackground};
  color: ${props => props.theme.colors.text};
  padding: 15px;
  padding-right: ${props => (props.withRightElement ? '85px' : '15px')};
  width: 100%;
  outline: none;
  font-family: ${props => props.theme.fontFamily};

  ::placeholder {
    opacity: 0.5;
  }
`;

const Wrapper = styled.div`
  position: relative;
`;

const RightElement = styled.div`
  position: absolute;
  top: 15px;
  right: 15px;
`;

const Input = getDefaultStyles('input');
const Textarea = getDefaultStyles('textarea');

type Props = {
  inputType?: 'input' | 'textarea',
  value: string,
  onChange?: string => void,
  onFocus?: (SyntheticFocusEvent<HTMLInputElement>) => void,
  rows?: number,
  disabled?: boolean,
  type?: string,
  step?: number,
  name?: string,
  renderRight?: () => Element<*> | null,
  bgColor?: string,
};

export const InputComponent = ({
  inputType,
  bgColor,
  onChange = () => {},
  renderRight = () => null,
  ...props
}: Props) => {
  const rightElement = renderRight();

  const inputTypes = {
    input: () => (
      <Input
        onChange={evt => onChange(evt.target.value)}
        withRightElement={Boolean(rightElement)}
        bgColor={bgColor}
        {...props}
      />
    ),
    textarea: () => (
      <Textarea onChange={evt => onChange(evt.target.value)} bgColor={bgColor} {...props} />
    ),
  };

  if (!Object.keys(inputTypes).find(key => key === inputType)) {
    throw new Error(`Invalid input type: ${String(inputType)}`);
  }

  return (
    <Wrapper>
      {inputTypes[inputType || 'input']()}
      {rightElement && <RightElement>{rightElement}</RightElement>}
    </Wrapper>
  );
};

InputComponent.defaultProps = {
  inputType: 'input',
  rows: 4,
  disabled: false,
  type: 'text',
  name: '',
  renderRight: () => null,
  onChange: () => {},
  onFocus: () => {},
  step: 1,
  bgColor: theme.colors.inputBackground,
};
