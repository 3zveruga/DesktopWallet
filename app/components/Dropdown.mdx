---
name: DropDown
---

import { Playground, PropsTable } from 'docz'

import { Dropdown } from './dropdown.js'
import { DoczWrapper } from '../theme.js'

# DropDown

<PropsTable of={Dropdown} />

## Basic usage

<Playground>
  <DoczWrapper>
    {() => (
      <div style={{ height: '500px' }}>
        <Dropdown
          options={[
            { label: 'asbh1yeasbdh23848asdasd', onClick: console.log },
            { label: 'urtyruhjr374hbfdjdhuh', onClick: console.log },
          ]}
          renderTrigger={toggleVisibility => <button onClick={toggleVisibility}>Opa</button>}
        />
      </div>
    )}
  </DoczWrapper>
</Playground>
