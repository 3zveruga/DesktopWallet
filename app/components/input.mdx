---
name: Input
---

import { Playground, PropsTable } from 'docz'

import { InputComponent } from './input.js'
import { DoczWrapper } from '../theme.js'

# Input

<PropsTable of={InputComponent} />

## Text Input

<Playground>
  <DoczWrapper>
    {() => <InputComponent inputType="input" value="Hello World!" onChange={console.log} />}
  </DoczWrapper>
</Playground>

## Textarea

<Playground>
  <DoczWrapper>
    {() => (
      <InputComponent
        inputType="textarea"
        value="I'm Zcash Electron Wallet"
        onChange={console.log}
        rows={10}
      />
    )}
  </DoczWrapper>
</Playground>
