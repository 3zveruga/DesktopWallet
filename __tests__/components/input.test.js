// @flow

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import 'jest-dom/extend-expect';

import { InputComponent } from '../../app/components/input';
import appTheme from '../../app/theme';

afterEach(cleanup);

describe('<InputComponent />', () => {
  describe('render()', () => {
    test('should render text input correctly', () => {
      const { container } = render(
        <ThemeProvider theme={appTheme}>
          <InputComponent
            inputType='input'
            value='Hello World!'
            onChange={console.log}
          />
        </ThemeProvider>
      );

      expect(container).toMatchSnapshot();
    });

    test('should render textarea correctly', () => {
      const { container } = render(
        <ThemeProvider theme={appTheme}>
          <InputComponent
            inputType='textarea'
            value='I am Zcash Electron Wallet'
            onChange={console.log}
            rows={10}
          />
        </ThemeProvider>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
