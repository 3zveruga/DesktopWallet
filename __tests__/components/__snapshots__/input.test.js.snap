// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputComponent /> render() should render text input correctly 1`] = `
<div>
  <div
    class="sc-bdVaJa EXKLM"
  >
    <input
      class="sc-htpNat cOoEZT"
      rows="4"
      step="1"
      type="text"
      value="Hello World!"
    />
  </div>
</div>
`;

exports[`<InputComponent /> render() should render textarea correctly 1`] = `
<div>
  <div
    class="sc-bdVaJa EXKLM"
  >
    <textarea
      class="sc-bxivhb klegai"
      rows="10"
      step="1"
      type="text"
    >
      I am Zcash Electron Wallet
    </textarea>
  </div>
</div>
`;
